packages/capacitor-plugin/android/src/main/java/com/capacitorjs/plugins/filetransfer/FileTransferPlugin.kt

```diff
@@ -0,0 +1,315 @@
+import android.os.Build
+import com.getcapacitor.JSObject
+import com.getcapacitor.PermissionState
+import com.getcapacitor.Plugin
+import com.getcapacitor.PluginCall
+import com.getcapacitor.PluginMethod
+import com.getcapacitor.annotation.CapacitorPlugin
+import com.getcapacitor.annotation.Permission
+import com.getcapacitor.annotation.PermissionCallback
+import io.ionic.libs.ionfiletransferlib.IONFLTRController
+import io.ionic.libs.ionfiletransferlib.model.IONFLTRDownloadOptions
+import io.ionic.libs.ionfiletransferlib.model.IONFLTRProgressStatus
+import io.ionic.libs.ionfiletransferlib.model.IONFLTRTransferHttpOptions
+import io.ionic.libs.ionfiletransferlib.model.IONFLTRTransferResult
+import io.ionic.libs.ionfiletransferlib.model.IONFLTRUploadOptions
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.flow.catch
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.onEach
+
+@CapacitorPlugin(
+    name = "FileTransfer",
+    permissions = [
+        Permission(
+            strings = [Manifest.permission.READ_EXTERNAL_STORAGE, Manifest.permission.WRITE_EXTERNAL_STORAGE],
+            alias = "publicStorage"
+        )
+    ]
+)
+class FileTransferPlugin : Plugin() {
+
+    companion object {
+        const val PUBLIC_STORAGE = "publicStorage"
+        private const val PROGRESS_UPDATE_INTERVAL = 100L // 100ms between progress updates
+    }
+
+    private val ioScope: CoroutineScope by lazy { CoroutineScope(Dispatchers.IO) }
+    private val controller: IONFLTRController by lazy { IONFLTRController(context) }
+    private lateinit var context: Context
+    private var lastProgressUpdate = 0L
+
+    override fun load() {
+        super.load()
+        context = bridge.context
+    }
+
+    override fun handleOnDestroy() {
+        super.handleOnDestroy()
+        ioScope.cancel()
+    }
+
+    private fun JSObject.toMap(): Map<String, String> {
+        val map = mutableMapOf<String, String>()
+        keys().forEach { key ->
+            map[key] = getString(key) ?: ""
+        }
+        return map
+    }
+
+    private fun JSObject.toParamsMap(): Map<String, Array<String>> {
+        val map = mutableMapOf<String, Array<String>>()
+        keys().forEach { key ->
+            val stringValue = getString(key)
+            if (stringValue != null) {
+                map[key] = arrayOf(stringValue)
+                return@forEach
+            }
+            
+            val arrayValue = getJSONArray(key)
+            if (arrayValue != null) {
+                val values = mutableListOf<String>()
+                for (i in 0 until arrayValue.length()) {
+                    arrayValue.optString(i)?.let { values.add(it) }
+                }
+                if (values.isNotEmpty()) {
+                    map[key] = values.toTypedArray()
+                }
+            }
+        }
+        return map
+    }
```

Comment:
Again, not super important, but this could be refactored to avoid nested if statements using a when block. Then, we could use the opt method directly to verify if the value is a String or a JSONArray, like so:

```
private fun JSObject.toParamsMap(): Map<String, Array<String>> {
    val map = mutableMapOf<String, Array<String>>()
    keys().forEach { key ->
        when (val value = opt(key)) {
            is String -> map[key] = arrayOf(value)
            is org.json.JSONArray -> {
                val values = Array(value.length()) { i -> value.optString(i) }
                if (values.isNotEmpty()) {
                    map[key] = values
                }
            }
        }
    }
    return map
}
```